nomenclature/codelist.py

```diff
@@ -542,3 +541,47 @@ def from_directory(cls, name: str, path: Path, file_glob_pattern: str = "**/*"):
             mapping[code.name] = code
 
         return cls(name=name, mapping=mapping)
+
+    def filter(self, hierarchy: str) -> "RegionCodeList":
+        """A filter that returns the components within a provided hierarchy.
+
+        Parameters
+        ----------
+        hierarchy : str
+            The name of the hierarchy whose components you are looking for.
+        Raises
+        ------
+        ValueError
+            If User inputs a typo or a hierarchy not compatible with the given model.
+        Returns
+        -------
+        RegionCodeList
+            Returns a list of the component regions of the inputted hierarchy.
+        """
+
+        mapping = {k: v for k, v in self.mapping.items() if v.hierarchy == hierarchy}
+        if mapping:
+            return RegionCodeList(name=self.name, mapping=mapping)
+        else:
```

Comment:
This is a nice error message and elegant code, but it is a lot of code and will be hard to maintain going forward. Better to simply raise: 
```python
f"Filtered RegionCodeList is empty: hierarchy={hierarchy}\nUse `RegionCodeList.hierarchy` for available items."
```

And then add a placeholder
```python
def hierarchy(self):
    raise NotImplementedError("This method is not yet implemented.")
```

And you can add this method as a future PR.